MVC

NicoMSK opened the pull request.

NicoMSK committed:
Make a function that calculates the current date

NicoMSK committed:
Create API functions and business models

-The current date display function has been created
-An API access function
has been created
-A class has been created for requesting a server

NicoMSK committed:
feat: Make a rendering of the main weather widget for today

- Make a drawing of the footer with weather data for today and tomorrow.
- Animate the flag for switching the temperature value from Celsius to Faringate

NicoMSK committed:
Make a selection of the current button

kagegrifon commented on package.json:
```diff
@@ -6,14 +6,16 @@
     "build": "tsc && vite build",
-    "preview": "vite preview"
+    "preview": "vite preview",
+    "test": "vitest"
```
о, круто)

kagegrifon commented on package.json:
```diff
@@ -6,14 +6,16 @@
     "dev": "vite",
     "build": "tsc && vite build",
-    "preview": "vite preview"
+    "preview": "vite preview",
```
а это зачем?

NicoMSK replied:
Не могу сказать) Я не помню, что б такое ставил)

kagegrifon commented on src/model/dateModel.ts:
```diff
@@ -0,0 +1,13 @@
+export function dateCurrent(date: Date) {
```
dateCurrent - функция должна быть глаголом

этот файл стоит переименовать. Это не модель, а всего ли утилита. Файл можно назвать именем функции

NicoMSK replied:
Переименовали

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,67 @@
+import * as temperatureToggle from "@/view/temperatureSwitching.view";
+
+const dateCurrent = date.dateCurrent(new Date());
+const celsiusFahrenheitSwitch = temperatureToggle.temperatureSwitching();
```
не очень понятно что происходит в этой строчке. Я имею ввиду по названием переменных не понятно

NicoMSK replied:
Удалили вообще и переменную и функуцию

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,67 @@
+async function startRenderWeather() {
+  const renderApiToday = weatherModel.getWeather();
+  const renderFooterToday = weatherModel.dataForFooterRender(
+    await renderApiToday,
```
а учитывается ли что renderApiToday может "упасть", т.е. зарезолвится с ошибкой?

kagegrifon replied:
обычно в коде так не пишут. Обычно async await используют, чтобы показать что вот тут мы ждем код, и что-то делаем дальше
```
  const renderApiToday = await weatherModel.getWeather(); // ждем пока выполнится getWeather()
  делаем что-то дальше после получения данных
```

когда ты разносишь await для renderApiToday, 
1) дублируется await и его легко забыть
2) как будто может быть ситуация, когда мы передадим renderApiToday без await - и это что-то очень необычное

renderApiToday - непонятное название для переменной. Скорее всего тут лежат данные по погоде?
тогда может weatherData?

kagegrifon replied:
мы помним, что может вернуться null?

NicoMSK replied:
> а учитывается ли что renderApiToday может "упасть", т.е. зарезолвится с ошибкой?

Я добавил сейчас try\catch.

NicoMSK replied:
> обычно в коде так не пишут. Обычно async await используют, чтобы показать что вот тут мы ждем код, и что-то делаем дальше
> 
> ```
>   const renderApiToday = await weatherModel.getWeather(); // ждем пока выполнится getWeather()
>   делаем что-то дальше после получения данных
> ```
> 
> когда ты разносишь await для renderApiToday,
> 
> 1. дублируется await и его легко забыть
> 2. как будто может быть ситуация, когда мы передадим renderApiToday без await - и это что-то очень необычное
> 
> renderApiToday - непонятное название для переменной. Скорее всего тут лежат данные по погоде? тогда может weatherData?

Поправил момент с await и переименовал переменную

NicoMSK replied:
> мы помним, что может вернуться null?

Вот этот вопрос не очень понял, давай разберём на созвоне.

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,67 @@
+    await renderApiToday,
+    dateCurrent,
+    celsiusFahrenheitSwitch,
+    "&deg;C"
```
такие штуки пишут в константы. И как будто это не нужно в контроллере, а должно быть в о view

NicoMSK replied:
Да, согласен, пришлось пока тут написать, так как функция не готова, а так да всё это во view будет

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,67 @@
+});
+
+renderView.tommorowButton?.addEventListener("click", async () => {
+  const weatherModelTomorrow = new WeatherModel("Moscow", "forecast", 2);
```
дублируются данные с Москвой с [этого места](https://github.com/NicoMSK/Weather-forecast-app/pull/2/files#diff-608df8b5b17aa28cc40e8633b2fa5f46912a15341d5fa46df3b31db15764c7c0R10) - напрашивается переменная

kagegrifon replied:
нам не обязательно каждый раз создавать новую модель new WeatherModel

обычно мы создаем модель и дальше ей пользуемся. По крайней мере если нам нужен один такой компонент

то, что тебе нужно - это добавить метод в WeatherModel для изменения данных, которые хочется поменять
changeLocation
changeForecast...

NicoMSK replied:
> дублируются данные с Москвой с [этого места](https://github.com/NicoMSK/Weather-forecast-app/pull/2/files#diff-608df8b5b17aa28cc40e8633b2fa5f46912a15341d5fa46df3b31db15764c7c0R10) - напрашивается переменная

Так это пока от руки написано, позже будут данные приходить из функции поиска по локации

NicoMSK replied:
> нам не обязательно каждый раз создавать новую модель new WeatherModel
> 
> обычно мы создаем модель и дальше ей пользуемся. По крайней мере если нам нужен один такой компонент
> 
> то, что тебе нужно - это добавить метод в WeatherModel для изменения данных, которые хочется поменять changeLocation changeForecast...

Поправил, убрал вообще конструктор и сделал метод, который принимает аргументы по локации

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,67 @@
+  const renderApiTomorrow = weatherModelTomorrow.getWeather();
+  const renderFooterTomorrow = weatherModelTomorrow.dataForFooterRender(
+    await renderApiTomorrow,
+    (currentDateMode = "tommorow")
```
(currentDateMode = "tommorow") - так делать не нужно
присвоение значения отдельно и передача параметра отдельно

NicoMSK replied:
Не знал, думал так компактней можно и присвоить и записать, ок, поправил.

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,67 @@
+    (currentDateMode = "tommorow")
+  );
+
+  renderView.mainWeather.renderFooter(await renderFooterTomorrow, "time", "℃");
```
не придется так делать когда будет const renderApiTomorrow = await weatherModelTomorrow.getWeather()

советую тебе писать без async await, так как похоже тебя этот синтаксис запутывает

NicoMSK replied:
Так тут нельзя без async await, не работает код без них. И я переделал на const weatherDataTomorrow = await weatherModel.getWeather(); Перенёс именно в переменную

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,67 @@
+});
+
+renderView.threeDaysButton?.addEventListener("click", async () => {
+  const weatherModelThreeDays = new WeatherModel("Moscow", "forecast", 3);
```
в обработчиках кнопок происходят одни и те же действия. Попахивает возможностью вынести это в функцию

NicoMSK replied:
Сделал одну общую функцию, чтоб не дублировать код, но работает на 2/3 кнопок. Давай на созвоне разберём как сделать более гибко и универсально.

kagegrifon commented on src/model/api/type.ts:
```diff
@@ -0,0 +1,139 @@
+type Condition = {
```
type.api.ts - можно убрать слово api, так как этот файл находится в папке api

NicoMSK replied:
Убрал.

kagegrifon commented on src/model/api/type.ts:
```diff
@@ -0,0 +1,139 @@
+};
+
+type Astro = {
+  sunrise: string;
```
круто, что все протипизировал!

kagegrifon commented on src/model/key.ts:
```diff
@@ -0,0 +1 @@
+export const API_KEY = "4a900778688045d887d103458252808";
```
уже описан в файле api. надо бы оставить что-то одно

NicoMSK replied:
Я вообще удалил этот файл. Он реально лишний

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,82 @@
+    let imgUrl = null;
+    let imgText = null;
+    let temperature = null;
+    let dataRenderDays = [];
```
а тут нужен let?

NicoMSK replied:
Вообще нет, поправил на const

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,82 @@
+    let imgText = null;
+    let temperature = null;
+    let dataRenderDays = [];
+    let hourseSort = null;
```
<img width="773" height="280" alt="image" src="https://github.com/user-attachments/assets/96e2f35d-1b8c-487f-a18c-44de29e98fde" />

kagegrifon replied:
хм?

NicoMSK replied:
Забавно, хз как так вышло, но переименовал)

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,82 @@
+    let dataRenderDays = [];
+    let hourseSort = null;
+
+    switch (timePeriod) {
```
это можно элегантно вынести в отдельный метод, getHourseSort()

 const hourseSort = this.getHourseSort(weatherData.forecast.forecastday);

NicoMSK replied:
Вот это не понял, что ты тут имеешь ввиду

kagegrifon replied:
имею ввиду вынести часть кода в отдельную функцию - метод класса

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,82 @@
+    weatherData: weatherType.ForecastDayWeather,
+    timePeriod: string
+  ) {
+    let time = null;
```
cудя по использованию, нет смысла объявлять эти переменные здесь
их нужно объявлять в цикле, где они и используются

NicoMSK replied:
Да, поправил, убрал из начала и объявил в цикле

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,82 @@
+      dataRenderDays.push({ time, imgUrl, imgText, temperature });
+    }
+
+    return dataRenderDays;
```
dataRenderDays -> renderDaysData

NicoMSK replied:
Переименовал и на 3 дня тоже

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  dataWeekForFooterRender(weatherData: weatherType.ForecastDayWeather) {
+    let date = null;
```
тоже нет смысла объявлять переменные заранее
если переменные за пределами цикла не используются, то они должны быть только внутри цикла
это облегчает чтение и понимание. Иначе создается впечатление, что они нужны еще где-то

NicoMSK replied:
Да, убрал из начала и объявил в цикле

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,82 @@
+
+    for (let i = 0; i < daysSort.length; i++) {
+      date = daysSort[i].date.split("-").reverse().join("-");
+      imgUrl = daysSort[i].day.condition.icon;
```
некоторые действия похожие со строками в dataForFooterRender. Тут можно вынести общую функцию, которая будет доставать данные в зависимости от типа режима
const curWeatherNormalizedData = getNormalizedWeatherData({ data: daysSort[i], type: 'daily'})
а для dataForFooterRender
const curWeatherNormalizedData = getNormalizedWeatherData({ data: evenHourse[i], type: 'hourly'})

NicoMSK replied:
Давай обсудим, потому что вроде понял, а вроде нет о чём ты.

kagegrifon commented on src/view/view.ts:
```diff
@@ -0,0 +1,160 @@
+  }
+}
+
+if (!(heroContainer instanceof HTMLElement)) {
```
а это нужно здесь?

NicoMSK replied:
Да, потому что TS ругается в этом месте: export const mainWeather = new MainWeather(heroContainer);

kagegrifon commented on src/view/view.ts:
```diff
@@ -0,0 +1,160 @@
+        </li>
+      </ul>`;
+
+    if (!(this.headerTime instanceof HTMLElement)) {
```
много уже подобных записей
лучше писать safetyQuery или ставить ! знаки. А то тяжко будет дальше

NicoMSK replied:
Что-то не нахожу инфы про safetyQuery. И почему тяжко будет дальше?

NicoMSK replied:
Если что я убрал instanceof HTMLElement и ошибок не появилось.

NicoMSK replied:
Если что я убрал instanceof HTMLElement и ошибок не появилось.

kagegrifon replied:
>Что-то не нахожу инфы про safetyQuery. И почему тяжко будет дальше?

мы это обсуждали в news - ты сам использовал эту функцию. Суть функции, что обработка ошибки находится внутри нее и снаружи не надо уже делать. Тяжко будет потому по коду нужно будет постоянно это проверять и это будут однообразные строчки кода, которые ухудшают читаемость

> Если что я убрал instanceof HTMLElement и ошибок не появилось.

тут стоит разобраться почему ты их тогда исходно добавлял. И почему ошибок не появилось. Хорошо, чтобы было ясно, а не как магия в духе: я не знаю почему, но ошибок не было. Тем более если раньше ошибки были

kagegrifon commented on src/view/view.ts:
```diff
@@ -0,0 +1,160 @@
+      10
+    );
+
+    this.renderFooter([], "", "");
```
[], "", "" - вот не понятно что происходит

чтобы такого не было лучше передавать параметры объектом

NicoMSK replied:
Переделал на объект

kagegrifon commented on src/view/view.ts:
```diff
@@ -0,0 +1,160 @@
+
+    this.parent.innerHTML = heroContent;
+    this.headerTime.innerHTML = weatherData.location.localtime.substring(
+      16,
```
уже встречал 16 и 10 раньше. Так что точно нужно выносить в константы и переиспользовать

NicoMSK replied:
В константы вынес, но как переиспользовать, если эти цифры находятся во view и model?

kagegrifon replied:
это ничего страшного. Надо понять происхождение (смысл) этих цифр. Т.е. что является причиной изменения этих цифр. И тогда будет понятно где их оставлять

возможно эти цифры будут как константы в одном месте и как входные параметры в другом

NicoMSK committed:
refactor: Made changes according to the commits from the code review

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,108 @@
+let currentDateMode: WeatherDateType = "today";
+const weatherModel = new WeatherModel();
+
+// async function startRenderWeather() {
```
коментики

NicoMSK replied:
убрал

kagegrifon commented on src/style/blocks/header.scss:
```diff
@@ -46,7 +46,7 @@
     position: absolute;
     bottom: -0.7rem;
-    width: 6rem;
+    width: 120%;
```
120% скорее всего приведет к непредсказуемому результату. Для чего тебе это понадобилось?
скорее всего тебе нужно что-то вроде calc(100% + 40px) или вообще по другому решить этот вопрос

NicoMSK replied:
поправил

NicoMSK committed:
refactor: Redo the render into one universal function for all three buttons

NicoMSK committed:
feat: Add a temperature switch from Celsius to Fahrenheit

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,97 @@
+    throw new Error("weatherDataToday не существует");
+  }
+
+  const renderFooterToday = await weatherModel.dataForFooterRender(
```
а нужно ли здесь тоже делать await? разве мы делаем несколько запросов в АПИ?

kagegrifon replied:
renderFooterToday - непонятно что это значит
<img width="739" height="311" alt="image" src="https://github.com/user-attachments/assets/c26d99c7-f23b-40b6-9f23-c2cce52c80e3" />

1) FooterToday -> TodayFooter. То что является существительным итоговым объектом должно быть на последнем месте. 
2) render - глагол

Что могло бы быть
- todayFooterNode (показываем что имеем дело с узлом dom дерева)
- todayFooter - просто показываем, что это футер (теряем контекст что это про ноды, но чем же еще может быть footer?)
- renderedTodayFooter - показываем, что это результат рендеринга 

судя по тому что это что-то из модели и по тому что это потом используется в renderFooter ниже, это не может быть нодами, это скорее всего какие-то данные
тогда видимо
- footerWeatherData
- footerItemsWeatherData
- detailedForecast
...

kagegrifon replied:
dataForFooterRender - вроде бы применяется как функция, а называется существительным
<img width="805" height="434" alt="image" src="https://github.com/user-attachments/assets/064f45c5-4d2c-447a-bb40-9587e6b4ea7f" />
пример такой
`const сидетьНаСтуле = await weatherModel.стул("синий")` - как понимать такой код?
вместо
`const результатОтПопыткиСесть = await weatherModel.сестьНаСтул('синий')`

NicoMSK replied:
Переименовал

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,97 @@
+    throw new Error(`${dataWeather} не существует`);
+  }
+
+  renderView.mainWeather.renderWeatherMain(
```
.mainWeather.renderWeatherMain

mainWeather  renderWeatherMain

mainWeather  WeatherMain

1) дублирование про weatherMain
2) измененный порядок слов ломает голову (и смысл). Это главная погода или погода главной?

NicoMSK replied:
переименовал

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,97 @@
+    weatherDateType.currentDateMode
+  );
+
+  const dateCurrent = date.formatDate(weatherDateType.dateDays);
```
dateCurrent - дата текущая
currentDate - текущая дата. В русском и английской обычно так говорят, а не наоборот

NicoMSK replied:
поправил

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,97 @@
+
+  const dateCurrent = date.formatDate(weatherDateType.dateDays);
+
+  const dataWeather = weatherModel.getDataForWeatherRender(
```
уже писал об этом много раз
dataWeather -> weatherData

NicoMSK replied:
переименовал

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,97 @@
+
+async function startRenderWeather(location: string) {
+  weatherModel.setParametersLocation(location, weatherDateType.days);
+  const weatherDataToday = await weatherModel.getWeather();
```
weatherDataToday  -> todayWeatherData
Что это за сущность? Это ДАННЫЕ. Поэтому это слово должно быть на последнем месте. Все остальные слова - это определение того, что это за данные. Данные чего? - Погоды. Какой погоды?  - Сегодняшней

а тут всегда про today будет? а tomorrow?

NicoMSK replied:
поправил

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,97 @@
+startRenderWeather("Moscow");
+
+temperatureToggle.temperatureToggle?.addEventListener("click", () => {
+  temperatureToggle.temperatureToggle?.classList.toggle("header__toggle--on");
```
контроллер не должен залезать в классы по хорошему
то как это происходит на совести view

kagegrifon replied:
мы можем синхронизировать модель и view

```
const currentUnit = weatherModel.toggleUnitTemperature();
view.temperatureToggler.setUnit(currentUnit)
```

тогда вьюшка просто слушает что возвращает модель и сама ничего не меняет, если вдруг по какой-то причине это не поменялось в модели. Т.е. если что-то сломается или надо поменять логику, то это всегда нужно делать в модели

NicoMSK replied:
Сделал отдельную функцию во view для переключения и синхронизировал view с model

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,97 @@
+renderView.todayButton?.addEventListener("click", async () => {
+  weatherDateType = {
+    currentDateMode: "today",
+    days: 1,
```
что значат поля days и dateDays? Имена непонятные

нет ощущения, что days, dateDays и currentDateMode непосредственно связаны?
currentDateMode: "today", -> days: 1, dateDays: 1,
currentDateMode: "tommorow", -> days: 2, dateDays: 0,
currentDateMode: "threeDay", -> days: 3, dateDays: 2,

если они непосредственно связаны, значит нам достаточно знать какой currentDateMode, а остальное мы можем вычислить, сделав объект соответствий
```
const DAY_OPT_BY_DATE_MODE = {
  today: { days: 1, dateDays: 0 },
  tommorow: { days: 2, dateDays: 1 },
  threeDay: { days: 3, dateDays: 2 },
}
```

NicoMSK replied:
переделали на объект

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,97 @@
+});
+
+renderView.threeDaysButton?.addEventListener("click", async () => {
+  weatherDateType = {
```
хорошо, у тебя получилось сделать общее состояние и менять его. Отлично!
теперь можно подумать как это отправить в модель)
менять данные напрямую не очень хорошо, так же как и давать доступ к такому - так как легко поломать

```
weatherModeModel.setMode('threeDay') - переключаем все параметры внутри модели

renderView.updateControlButtons(weatherModeModel.getCurrentMode().currentDateType); - даем только на чтение
```

NicoMSK replied:
не понял коммент, надо разобрать

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,145 @@
+const END_STRING = 10;
+
+export type WeatherDateType = {
+  currentDateMode: "today" | "tommorow" | "threeDay";
```
такой тип будет плохо соответствовать действительности, ведь у тебя не может быть "tommorow" с currentPeriod "date", правильно?
а такая типизация это позволяет
в данном случае удобнее описать похожим объектом, который я уже писал тебе выше, и применить магию из typescript)

```
const OPT_BY_DATE_MODE = {
  today: { days: 1, dateDays: 0, currentPeriod: "time" },
  tommorow: { days: 2, dateDays: 1, currentPeriod: "time" },
  threeDay: { days: 3, dateDays: 2, currentPeriod: "date" },
} as const;

export type WeatherDateMode = keyof typeof OPT_BY_DATE_MODE;
export type WeatherDateModeValues = typeof OPT_BY_DATE_MODE[WeatherDateMode]
```
посмотри что за типы будут получаться

NicoMSK replied:
поправил и сделал единую типизацию на основе: OPT_BY_DATE_MODE

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,145 @@
+export class WeatherModel {
+  location: string = "";
+  days: number = 0;
+  unit: "c" | "f" = "c";
```
это удобнее тоже описать константами

```
const TEMPERATURE_TYPE = {
  celcium: '℃',
   farenheight: '℉',
} as const

type TemperatureType = keyof typeof TEMPERATURE_TYPE

export class WeatherModel {
  ...
  unit: TemperatureType = "celcium";
  
  get unitSymbolTemperature(): { - тип будет из return
     return TEMPERATURE_TYPE[this.unit]
  }
```

NicoMSK replied:
нельзя указать словами какие градусы, максимум можно вот так сделать: 
`const TEMPERATURE_TYPE = {
  c: "℃",
  f: "℉",
} as const;`

А если указывать словами, то код не корректный и цифры не рендарятся

NicoMSK replied:
переписал через эту константу

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,145 @@
+    weatherData: weatherType.ForecastDayWeather,
+    period: string
+  ) {
+    const tempKeyDayCurrent = `temp_${this.unit}` as "temp_c" | "temp_f";
```
as "temp_c" | "temp_f"; 
не нужная и не удобная типизация, она не убережет от ошибки(

лучше написать утилиту 
```
function getTempKeyDayCurrent(unit) {
  if (unit === 'celcium') return "temp_c"
  if (unit === 'farenheight') return "temp_f"
  throw new Error('unexpected unit: ' + unit)
}
...
const tempKeyDayCurrent = getTempKeyDayCurrent(this.unit)
```
другой вариант 
`const tempKeyDayCurrent = this.unit === 'celcium' ? "temp_c" :  "temp_f"`

причем функцию можно объединить для всех ключей зависящих от типа температуры
```
function getTempKeyDayCurrent(unit) {
  if (unit === 'celcium') return { current: "temp_c", avg: 'avgtemp_c' }
  if (unit === 'farenheight') return { current: "temp_f", avg: 'avgtemp_f }
  throw new Error('unexpected unit: ' + unit)
}
```

kagegrifon replied:
опять с именем беда
tempKeyDayCurrent 
current тут лишний совсем (по смыслу)
day тоже лишний - потому что ключ не привязан к какому-то дню
Что здесь является сущностью? Видимо КЛЮЧ. Какой ключ? от температуры
итого tempKey 
но лучше не сокращать, так как temp - это часто сокращение слова temporary (временный), наверное замечал подобный папки и файлы
поэтому лучше писать полностью
temperaruteKey

NicoMSK replied:
Добавил утилиту и переименовал переменную

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,145 @@
+    switch (period) {
+      case "today":
+        return {
+          location: weatherData.location.name,
```
тут можно обратить внимание на те поля которые не отличаются при смене: location и localtime

остальное лучше вынести в отдельную функцию
```
return ({
  location: weatherData.location.name,
  localtime: weatherData.location.localtime,
  ...getFormattedDataFromApi(period) // работу с ключами тоже можно туда
})
```

NicoMSK replied:
Переписал и сделал отдельный метод

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,145 @@
+      case "tommorow":
+        hoursSort = weatherData.forecast.forecastday[1].hour;
+        break;
+      case "threeDay":
```
не хватает пробелов между строками
рекомендую использовать if else, так как он более читаем, меньше строк и для разных языков программирования выглядит почти одинаково.

NicoMSK replied:
разделил вообще на 3 метода и убрал switch

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,145 @@
+        hoursSort = weatherData.forecast.forecastday[1].hour;
+        break;
+      case "threeDay":
+        const daysSort = weatherData.forecast.forecastday;
```
daysSort -> дневная сортировка или сортировка дней
ты это имел ввиду?

NicoMSK replied:
Вообще daysSort --- сортировка по дням, вот так мне переводчик переводит

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,145 @@
+        return renderDaysData;
+    }
+
+    if (!hoursSort) throw new Error(`${hoursSort} не существует`);
```
лучше уж в самом начале в одном месте проверить на все несуществующее, чем раскидывать в разных местах кода

NicoMSK replied:
А как это в самом начале сделать, что-то не очень понял?

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,145 @@
+
+    if (!hoursSort) throw new Error(`${hoursSort} не существует`);
+
+    const evenHourse = hoursSort.filter((hour, index) => index % 2 === 0);
```
почему evenHourse, а не evenHour
Hourse я упорно хочу прочитать как лошадь))

NicoMSK replied:
Хороший вопрос, но я поправил на evenHours

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,145 @@
+        if (!daysSort) throw new Error(`${daysSort} не существует`);
+
+        for (let i = 0; i < daysSort.length; i++) {
+          const date = daysSort[i].date.split("-").reverse().join("-");
```
это все можно вынести в функцию makeDayForecast()
```
for (let i = 0; i < daysSort.length; i++) {
  const curDayForecast = makeDayForecast(daysSort[i])
  renderDaysData.push(curDayForecast);
}
```

NicoMSK replied:
я разделил на 3 метода и вынес эти циклы в отдельные 2 метода

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,145 @@
+    const evenHourse = hoursSort.filter((hour, index) => index % 2 === 0);
+
+    for (let i = 0; i < evenHourse.length; i++) {
+      const tempKeyDay = `temp_${this.unit}` as "temp_c" | "temp_f";
```
тоже можно вынести в отдельную функцию см про makeDayForecast

NicoMSK replied:
выше ответ

kagegrifon commented on src/view/view.ts:
```diff
@@ -0,0 +1,168 @@
+  renderWeatherMain(
+    weatherData: WeatherDataDay,
+    dateCurrent: string,
+    unitMeasurement: "℃" | "℉"
```
эти типы мы можем брать из модели

NicoMSK replied:
А как, если view не должна знать о model?

kagegrifon replied:
типы можно передавать
мы можем вынести типы на уровень выше модели и view - тогда можно)
как это общие типы))

NicoMSK committed:
refactor: Make changes according to the comments from the code review

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -1,97 +1,74 @@
-  if (!weatherDataToday) {
+  const currentOption = DAY_OPT_BY_DATE_MODE[currentDateMode];
+
+  weatherModel.setParametersLocation(location, currentOption.days);
```
setParametersLocation(location, currentOption.days); - тут передаешь и дни и локацию.
эти данные не связанны, не обязательно привязывать их к одному методу.
```
weatherModel.setLocation(location)
weatherModel.setDays(currentOption.days)
```
можно сделать и универсальный метод для изменения параметров
```
weatherModel.changeParameter({ location })
weatherModel.changeParameter({ location, days: currentOption.days }) // т.е. можно передавать сколько угодно изменений за раз
```
тебе для ясности предлагаю использовать отдельные

нужно ли на каждом рендере делать изменение локации и days?
разве это не должно происходить только при определенных переключениях?

kagegrifon replied:
currentOption.days скорее всего не нужен, если есть currentDateMode

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -1,97 +1,74 @@
+
+  weatherModel.setParametersLocation(location, currentOption.days);
+
+  const todayWeatherData = await weatherModel.getWeather();
```
чтобы не забыть
скорее всего нам не понадобиться на каждый рендер дергать бекенд
но мы не будем сейчас это разбирать в этом МР

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -1,97 +1,74 @@
-    footerItems: renderFooterToday,
-    period: weatherDateType.currentPeriod,
+    footerItems: footerWeatherData,
+    period: currentOption.currentPeriod as "time" | "date",
```
тут видимо проблема с типизацией, раз делается  as "time" | "date",
наверняка можно сделать без этого

as  - выключает проверку typescript и потенциально это место может сломаться

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -1,97 +1,74 @@
-
+renderView.todayButton?.addEventListener("click", () => {
+  currentDateMode = "today";
+  renderView.updateControlButtons(currentDateMode);
```
в этом месте и нужно менять состояние модели относительно режима currentDateMode

kagegrifon commented on src/model/model.ts:
```diff
@@ -31,115 +37,144 @@ export class WeatherModel {
+        precip: weatherData.current.precip_mm,
+        wind: weatherData.current.wind_kph,
+      };
     } else {
```
после if с return не обязательно ставить else, так как код точно дальше не пойдет - так легче читать

kagegrifon commented on src/model/model.ts:
```diff
@@ -31,115 +37,144 @@ export class WeatherModel {
-      return "℉";
+      const tempKeyThreeDay = this.getTempKeyDayCurrent(this.unit);
+
+      if (days === undefined) {
```
а разве у нас может быть days undefined?

NicoMSK replied:
убрал эту проверку

kagegrifon commented on src/model/model.ts:
```diff
@@ -31,115 +37,144 @@ export class WeatherModel {
+      };
     } else {
-      return "℉";
+      const tempKeyThreeDay = this.getTempKeyDayCurrent(this.unit);
```
повторение строчки 72

NicoMSK replied:
убрал вторую и сделал всё на одной

kagegrifon commented on src/model/model.ts:
```diff
@@ -31,115 +37,144 @@ export class WeatherModel {
+      }
+
+      return {
+        location: LOCATION,
```
а tomorrow где обрабатывается?

kagegrifon commented on src/model/model.ts:
```diff
@@ -31,115 +37,144 @@ export class WeatherModel {
-          precip: weatherData.forecast.forecastday[0].day.totalprecip_mm,
-          wind: weatherData.forecast.forecastday[0].day.maxwind_kph,
-        };
+        return this.getFormattedDataFromApi(weatherData, period, 0);
```
(weatherData, period, 0); - такие параметры плохо читаются, лучше делать объектом
при том еще путает, что в одно месте кода называется weatherDateMode, а тут period. Лучше везде одинаково делать

kagegrifon replied:
цифры 0 и 2 прибиты к period и мы их можем получить из объекта DAY_OPT_BY_DATE_MODE 

этот метод getDataForWeatherRender не делает ничего полезного, его можно убрать и использовать getFormattedDataFromApi

kagegrifon commented on src/model/model.ts:
```diff
@@ -31,115 +37,144 @@ export class WeatherModel {
+    period: string,
+    days?: number
+  ) {
+    const LOCATION = weatherData.location.name;
```
а почему названо БОЛЬШИМИ буквами? ведь это не константа

NicoMSK replied:
исправил

kagegrifon commented on src/model/model.ts:
```diff
@@ -31,115 +37,144 @@ export class WeatherModel {
-      const temperature = Math.round(evenHourse[i][tempKeyDay]);
+    for (let i = 0; i < evenHours.length; i++) {
+      renderDaysData.push({
+        time: getTextSlice(evenHours[i].time),
```
getTextSlice -> getFormattedTime 
getTextSlice  - получить кусочек текста
getFormattedTime - получить отформатированное время

NicoMSK replied:
окей, переименовал

kagegrifon commented on src/model/model.ts:
```diff
@@ -31,115 +37,144 @@ export class WeatherModel {
   }
+
+  getDataRenderFooter(
+    weatherData: weatherType.ForecastDayWeather,
```
а зачем нам получать weatherData как параметр, если мы ее можем сами получить из модели?

NicoMSK replied:
Переделал, убрал из параметров и сделал в методах через переменную диструктаризации

kagegrifon commented on src/util.ts:
```diff
@@ -0,0 +1,22 @@
+const END_STRING = 16;
+
+export function getTextSlice(string: string) {
+  return string.slice(START_STRING, END_STRING);
```
если START_STRING, END_STRING используются только в одном месте то можно и без выноса в переменные. Их имена не дают дополнительной информации

NicoMSK replied:
убрал переменные и просто вписал цифры

kagegrifon commented on src/util.ts:
```diff
@@ -0,0 +1,22 @@
+  return string.slice(START_STRING, END_STRING);
+}
+
+export function safetyQuerySelector<T extends Element>(
```
о, молодец!!
почувствовал как код стал чище?

kagegrifon replied:
с гпт написал или сам?)

NicoMSK replied:
> о, молодец!! почувствовал как код стал чище?

я бы сказал не столько чище, сколько нагруженный для чтения)

NicoMSK replied:
> с гпт написал или сам?)

сам) просто за основу взял из проекта про новости тот код и допили под этот проект

NicoMSK committed:
refactor: Make changes according to the comments from the code review

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -5,34 +5,17 @@ import * as renderView from "@/view/view";
-    todayWeatherData,
-    currentDateMode
-  );
+  await weatherModel.getWeather();
```
этот метод может возвращать данные в уже нужном формате (внутри себя будет возращать через getFormattedDataFromApi) и тогда не нужно специально дергать getFormattedDataFromApi ниже
`const weatherData = await weatherModel.getWeather()`

NicoMSK replied:
Ок, переправил, но не понимаю для чего так делать.

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -42,33 +25,33 @@ async function startRenderWeather(location: string) {
-startRenderWeather("Moscow");
+startRenderWeather();
 
 temperatureSwitch.temperatureToggle?.addEventListener("click", () => {
```
вопросик больше не нужен, так как мы получаем через safetyQuery

NicoMSK replied:
Если убрать "?", то TS ругается, так как temperatureSwitch.temperatureToggle может быть null, поэтому убрать нельзя

kagegrifon commented on src/model/model.ts:
```diff
@@ -51,84 +65,60 @@ export class WeatherModel {
-  ) {
-    const LOCATION = weatherData.location.name;
-    const LOCAL_TIME = weatherData.location.localtime;
+  getFormattedDataFromApi() {
```
я бы предложил передавать weatherData в параметрах (и требовать параметрах. чтобы она не была null, а только weatherType.ForecastDayWeather
`getFormattedDataFromApi(weatherData: weatherType.ForecastDayWeather) {`

чтобы потом не нужно было this.validateWeatherData();

validateWeatherData - не очевидно что делает эта функция, более читаемо было бы 
if (!this.data) return / throw - сразу видно что происходит. А от валидации ожидания разные

ведь метод getFormattedDataFromApi не имеет смысла без данных, верно? а значит он жестко от них зависит

NicoMSK replied:
хорошо, передал weatherData на параметр

kagegrifon commented on src/model/model.ts:
```diff
@@ -167,14 +159,11 @@ export class WeatherModel {
-      return this.getWeatherForThreeDay(weatherData);
+  getDataRenderFooter() {
+    if (this.currentDateMode === "threeDay") {
+      return this.getWeatherForThreeDay();
```
чтобы не нужно было validateData делать внутри getWeatherForThreeDay и getWeatherForDay
эту проверку можно сделать тут и в тоже время предусмотреть случай, если null
```
getDataRenderFooter() {
  if (!this.weatherData) return null
  
  // дальше спокойно работаем и вызываем методы передавая в параметры 
  return this.getWeatherForThreeDay(weatherData);
  

}
```
т.е. для тех методов, которые как-то преобразовывают данные, не стоит предусматривать случай, что этих данных нет - это лучше предусмотреть логикой выше. В данном случае лучше один раз сделать проверку в getDataRenderFooter, чем во всех функциях ниже: getWeatherForThreeDay, getWeatherForDay. Даже если бы внутри была одна функция, эту логику лучше предусмотреть выше - так как она определяет, что вообще нужно делать в случае отсутствия данных, а конкретные функции ниже просто преобразовывают данные и не задумываются об остальном - разделение ответственности и обязанностей

NicoMSK replied:
ок, убрал проверки из getWeatherForThreeDay и getWeatherForDay и сделал одну в getDataRenderFooter

kagegrifon commented on src/model/model.ts:
```diff
@@ -1,6 +1,6 @@
 import * as weatherApi from "@/model/api/weather.api.ts";
```
давно хотел сказать, что этот файл лучше переименовать в weatherModel)
это не критично

NicoMSK replied:
А почему именно в weatherModel переименовать, если это API?

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,169 @@
+
+export class WeatherModel {
+  location: string = "Moscow";
+  days: number = 0;
```
уже не нужно

NicoMSK replied:
Говорили же, что позже при подключении API по городам, это пригодится или всё-таки нет?

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,169 @@
+    this.days = days;
+  }
+
+  validateWeatherData(): asserts this is {
```
эту функцию можно было бы вынести из класса и передавать данные через аргумент

и в тоже время можно обойтись без нее, потому что останется только 1 место где нужно будет проверить - в getDataRenderFooter

getFormattedDataFromApi - можно вызывать внутри getWeather 
а для getWeatherForDay и getWeatherForThreeDay будет проверяться выше в getDataRenderFooter

NicoMSK replied:
ок, убрал её вообще

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,169 @@
+  async getWeather() {
+    const daysOpt = DAY_OPT_BY_DATE_MODE[this.currentDateMode].days;
+    try {
+      return (this.weatherData = await weatherApi.getWeatherFromAPI(
```
так делать не красиво))

не нужно на return делать присвоение - это плохо читаемо и вызывает недоумение - нет ли тут ошибки. 
```
this.weatherData = await weatherApi.getWeatherFromAPI(this.location,daysOpt)

return this.weatherData
```

NicoMSK replied:
Переделал вообще, чтоб: return this.getFormattedDataFromApi();

NicoMSK committed:
feat: Add a download window and error message display.

- Corrected the rest of the code according to the comments from the code review

kagegrifon commented on src/controller/controller.ts:
```diff
@@ -0,0 +1,62 @@
+async function startRenderWeather() {
+  const currentOption = DAY_OPT_BY_DATE_MODE[weatherModel.currentDateMode];
+
+  const weatherData = await showDownload.showsDownloadWindow(
```
showDownload.showsDownloadWindow -> loader.show()

обычно делают так
```
loader.show()
const weatherData = await weatherModel.getWeather()
loader.hide()
```

а если хочешь дополнительно обработать случай с null, то можно 
```
loader.show()
const weatherData = await weatherModel.getWeather()
loader.hide()

if (!weatherData) {
  // тут показываем сообщение об ошибке и предложении обновить страницу и прочее
  return // чтобы не выполнять дальнейший код
} 
```
если оставляем throw Error в weatherModel.getWeather
```
loader.show()
let weatherData
try {
  weatherData = await weatherModel.getWeather()

} catch (e) { // если генерируем кастомные, то можем использовать и смотреть что за ошибка (сети или сервера)
     // тут показываем сообщение об ошибке и предложении обновить страницу и прочее
    return  // чтобы не выполнять дальнейший код
} finally() {
    loader.hide()
}

```

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,165 @@
+  }
+
+  getFormattedDataFromApi(weatherData: weatherType.ForecastDayWeather) {
+    if (!this.weatherData) {
```
это лишнее тогда

NicoMSK replied:
Убрал

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,165 @@
+      hoursSort = forecast.forecastday[1].hour;
+    }
+
+    if (!hoursSort) throw new Error("hoursSort не существует");
```
это не нужно (попробуй удалить)

NicoMSK replied:
Удалил и реально, не заругалось, но вопрос почему до этого ругалось

kagegrifon commented on src/view/downloadWindow.ts:
```diff
@@ -0,0 +1,26 @@
+const spinnerLoading = safetyQuerySelector<HTMLElement>(".dialog__loader");
+
+export async function showsDownloadWindow(promise: Promise<T>) {
+  const timeoutId = setTimeout(() => loadingData.showModal(), 150);
```
а почему на 150 милисекунд? а не на 300 или 1000

так делать неправильно
промис может произвольно долго выполняться
тут уж тогда надо было от промиса отталкиваться

```
loadingData.showModal()
return promise
  .then(() =>  loadingData.close())
  .catch(() => {
    textLoadingStatus.textContent =
      "Ошибка загрузки данных. Попробуйте ещё раз.";
    spinnerLoading.style.display = "none";
  })

```

kagegrifon commented on src/view/view.ts:
```diff
@@ -0,0 +1,167 @@
+        </li>
+      </ul>`;
+
+    if (!this.headerTime) {
```
не нужно

NicoMSK replied:
удалил

kagegrifon commented on src/view/view.ts:
```diff
@@ -0,0 +1,167 @@
+    period: "time" | "date";
+    unitMeasurement: "℃" | "℉";
+  }) {
+    if (!(this.weatherClockList instanceof HTMLElement)) {
```
тоже не нужно.
для этого и использовали safetyQuery

NicoMSK replied:
согласен, убрал

kagegrifon commented on src/model/model.ts:
```diff
@@ -0,0 +1,165 @@
+      daysOpt
+    );
+
+    if (!this.weatherData) {
```
можно и так
```
if (this.weatherData) return this.getFormattedDataFromApi(this.weatherData);

return null
```
но тогда тебе надо будет переделать loader
 в случае ошибки сети getWeatherFromAPI (см catch) возвращает null - и это можно тоже обрабатывать  - показать сообщение как это у тебя уже делается - а можно и по разному обрабатывать (когда допустим null)

внешний обработчик ошибки промиса будет триггериться на ошибку по статусу не ок
```
    if (!response.ok) {
      throw new Error(`Запрос вернулся с ошибкой ${response.status}`);
    }
```

в целом можешь и так оставить как у тебя есть, так как у тебя сейчас лоадер перехватывает ошибки и обрабатывает

kagegrifon replied:
есть еще подходы по работе с ошибками - генерация разных видов ошибок и их обработка, но давай тут не будем усложнять

kagegrifon replied:
https://learn.javascript.ru/custom-errors
https://doka.guide/js/errors/
https://habr.com/ru/companies/reksoft/articles/518050/ - сложная

NicoMSK replied:
> в целом можешь и так оставить как у тебя есть, так как у тебя сейчас лоадер перехватывает ошибки и обрабатывает

В этом и была задумка, потому что модалка с сообщением ошибки не хотела появляться.

NicoMSK committed:
refactor: Make edits according to the comments from the codeview
